Kabanov.  Task №4

kswiftie opened the pull request:
# Задача 4. Пул потоков

* **Дедлайн**: 24.10.2024, 23:59
* Полный балл: 10

## Задача

- [x] Реализовать паттерн "пул потоков".
  - [x] Число потоков задаётся константой в классе пула или параметром конструктора.
  - [x] У каждого потока есть два состояния: ожидание задачи, выполнение задачи.
  - [x] Задача --- некая функция проводящая вычисления и возвращающая результат.
  - [x] Добавление задачи осуществляется с помощью метода `enqueue(task)`
  - [x] При добавлении задачи, если в пуле есть ожидающий поток, то он должен приступить к её выполнению. Иначе задача будет ожидать исполнения, пока не освободится какой-нибудь поток.
  - [x] Метод `dispose` должен завершать работу потоков. Уже запущенные задачи не прерываются, но новые задачи не принимаются на исполнение потоками из пула.
  - [x] Ограничения:
    - Разрешено использовать только пакет threading
    - Для каждого базового сценария должен быть написан несложный тест
    - Также должен быть написан тест, проверяющий, что в пуле действительно не менее `n` потоков
  - [x] Допущения:
    - Использование класса для реализации пула потоков необязательно
- [x] **(Доп., +2 балла)** С помощью `concurrent.futures` реализовать параллельное вычисление суммы полного декартового произведения множества целых чисел

kswiftie committed:
1. Added module threadpool with all functional

2. Added tests for module threadpool, tests.utils.py have been updated

kswiftie committed:
Fixes for requirements.txt and github workflows

kswiftie committed:
Fixes for pre-cimmit-config.yaml and for code style

kswiftie committed:
code style fixed

Krekep commented on project/threadpool/threadpool.py:
```diff
@@ -0,0 +1,87 @@
+        """
+        while not self.is_threads_opened:
+            self.event.wait()
+            while self.tasks:
+                try:
+                    task, args, kwargs = self.tasks.popleft()
+                    result = task(*args, **kwargs)
+                except IndexError:
+                    break
```
Зачем так сложно? По факту же это не цикл, а `if`, нет?

kswiftie replied:
использование try except здесь бесполезно, while not self.is_threads_opened тоже не нужно, но нужен while self.tasks
Если у меня будет структура:
```
    while not self.is_threads_opened:
        self.event.wait()
        if self.tasks:
            task, args, kwargs = self.tasks.popleft()
            result = task(*args, **kwargs)
```
то задачи будут будут выполняться пока потоки открыты, но при вызове Threadpool.dispose(), задачи, не успевшие начать обрабатываться, завершены не будут

kswiftie replied:
Поэтому я исправил runner так:
```
       self.event.wait()
        while self.tasks:
            task, args, kwargs = self.tasks.popleft()
            result = task(*args, **kwargs)
        self.event.clear()
```
когда добавляются задачи, потоки начинают их искать, и считают, пока не выполнят все. Потом self.event снова переходит в режим ожидания до добавления новых задач. Threadpool.dispose() будет ждать, пока завершаться задачи, не успевшие обработаться

Krekep commented on tests/test_threadpool.py:
```diff
@@ -0,0 +1,146 @@
+            5,
+        ),
+        (
+            Threadpool(5),
+            [wait_signal for _ in range(12)],
+            [threading.Event() for _ in range(12)],
+            list(range(4)),
+            3,
+            5,
```
Если я правильно понимаю, то Вы добавляете 12 задач, затем 4 (`(list(range(4))`) завершаете, верно? Почему тогда в очереди остаётся 3 задачи?
Или я неверно понял тест

kswiftie replied:
Когда я добавляю 12 задач, 5 начинают обрабатываться, 7 в очереди. Я завершил 4, теперь 4 перешло в обработку, в ожидании висят 3 задачи

Krekep commented on tests/test_threadpool.py:
```diff
@@ -0,0 +1,146 @@
+        ),
+    ],
+)
+def test_runner(threadpool_model, func, inputs, expected):
+    to_remember = set()
+    for inp in inputs:
+        threadpool_model.enqueue(func, inp, st=to_remember)
+
+    threadpool_model.dispose()
+    assert to_remember == expected
```
В этом тесте несколько потоков добавляют объекты в одно и тоже множество без блокировок, верно?
Если так, то выглядит очень ненадёжно и непотокобезопасно

kswiftie replied:
Верно. Исправил

Krekep commented on tests/test_threadpool_cartesia_product.py:
```diff
@@ -0,0 +1,29 @@
+        (get_function_execution_time, [cartesian_sum, set(range(1, 6)), 12], 0.5),
+    ],
+)
+def test_cartesian_sum_time(function, args, expected_time):
```
Ну здесь как-будто и однопоточная версия будет достаточно быстрой, так что тест не особо презентативный

kswiftie replied:
Я сейчас заметил, что cartesian_sum работает ужасно медленно, сильно медленнее однопоточного вычисления, я долго пытался оптимизировать это, но у меня ничего не получилось, и сейчас у меня нет идей. Вы можете подсказать, как сделать это быстро? Или это не обязано работать быстрее однопоточного вычисления? У меня возникла проблема с тем, что для многопоточности мне нужно разбить исходные данные на независимые части, и, видимо, именно это занимает много времени

Krekep replied:
Хммм, вообще с использованием ProcessPool должно быть быстрее, но стоит помнить, что этот выигрыш достигается на больших множествах, а на маленьких однопоточный вариант будет быстрее, за счёт отсутствия оверхеда на многопроцессорность.
Можете почитать [вот этот тред](https://stackoverflow.com/questions/42074501/python-concurrent-futures-processpoolexecutor-performance-of-submit-vs-map)

Krekep replied:
Ладно, скорее всего я был неправ, давая такое задание. Для более тяжёлых функций, чем `sum` задание имело бы смысл, но в с учётом того, что `sum` использует сишную реализацию, оверхед на копирование массивов между процессами и их синхронизацию сильно больше, чем выигрыш от их использования в такой простой задаче. Так что окей, можете не добавлять такой тест. Либо можете добавить более тяжёловесную функцию (добавить, к примеру косинусы, синусы, корни или, к примеру, сортировку сделать) и показать, что с процессами она будет быстрее
Также можете почитать [этот (другой) тред](https://www.reddit.com/r/Python/comments/5tfm64/multiprocessing_parallel_sum_slower_than_serial/)

kswiftie replied:
Я убрал этот тест. Я пытался добавить новую функцию, но она как-то несущественно быстрее однопоточной работала и у нее терялась точность, так что я не придумал, что можно добавить вместо суммы прямого произведения для демонстрации скорости от этого модуля

Krekep replied:
К примеру
```
def single_func(a: list) -> int:
    res = 0
    for elem in a:
        res += math.sqrt(math.cos(elem) * math.sin(elem) + 1)
    return res
```

На 4 процессах работает примерно в 6 (это много) раз быстрее, чем её однопоточный вариант. Единственное --- я преобразовал массив кортежей в одномерный массив

Krekep reviewed (changes requested):
Необходимы исправления/комментарии

kswiftie committed:
1. Updated cartesiad_product_sum

2. Updated class Threadpool

3. Updated utils and removed test for cartesian sum

Krekep commented on project/threadpool/threadpool.py:
```diff
@@ -0,0 +1,83 @@
+
+
+        """
+        self.event.wait()
+        while self.tasks:
+            task, args, kwargs = self.tasks.popleft()
+            result = task(*args, **kwargs)
+        self.event.clear()
```
Представим, что у нас `n` потоков, пришло `n` задач, потом перерыв на несколько секунд (во время которых потоки закончат работу и соответственно функцию `runner`) и снова `n` задач. Без `while True` разве последние `n` выполнятся?

kswiftie replied:
Нужен, но не while True, а while not self.is_threads_closed, чтобы после вызова dispose выполнились оставшиеся задачи, и потоки завершили работу. Исправил, добавил тесты, демонстрирующие это

Krekep reviewed (changes requested):
Всё хорошо, но момент про функцию `runner` требует уточнения. Возможно я не умею в голове компилировать программы, но тогда добавьте соответствующий тест

kswiftie committed:
Fixes for Threadpool and new tests for it

Krekep reviewed (approved):
Зачтено
